File: drills/elixir/solutions/nc138_diameter_of_binary_tree__height_per_node.exs
defmodule Solution do
  # Ask every node how tall its two sides are and keep the largest sum. Correct
  # and obvious, but height is recomputed from scratch at every node, so a
  # balanced tree costs O(n log n) and a spindly one O(n^2) -- which is exactly
  # what returning the height alongside the answer avoids.
  def diameter_of_binary_tree(nil), do: 0

  def diameter_of_binary_tree({_value, left, right}) do
    Enum.max([
      height(left) + height(right),
      diameter_of_binary_tree(left),
      diameter_of_binary_tree(right)
    ])
  end

  defp height(nil), do: 0
  defp height({_value, left, right}), do: 1 + max(height(left), height(right))
end

File: drills/go/solutions/nc138_diameter_of_binary_tree__height_per_node.go
package main

func diameterOfBinaryTree(root *TreeNode) int {
	// For each node, measure both subtree heights from scratch and keep
	// the best: O(n^2) on a skewed tree, but a direct reading of the
	// definition.
	if root == nil {
		return 0
	}
	through := height(root.Left) + height(root.Right)
	return max(through, max(diameterOfBinaryTree(root.Left), diameterOfBinaryTree(root.Right)))
}

func height(node *TreeNode) int {
	if node == nil {
		return 0
	}
	return 1 + max(height(node.Left), height(node.Right))
}
